Docs: Story globals improvements (#28856)
storybookjs/storybook: docs/essentials/backgrounds.mdx

<Callout variant="warning" icon="🆕" id="globals-api">
  **`globals` API**

  This addon has been updated to use the `globals` API when the [`backgroundsStoryGlobals` feature flag](../api/main-config/main-config-features.mdx#backgroundsstoryglobals) is enabled. With `globals`, when you specify a background value for a story, it cannot be overridden from the toolbar. This ensures a consistent experience while navigating between stories. This will be the default behavior and API in Storybook 9.
</Callout>

## Configuration

By default, the backgrounds toolbar includes a light and dark background.

But you're not restricted to these backgrounds. You can configure your own set of colors with the `backgrounds` [parameter](../writing-stories/parameters.mdx) in your [`.storybook/preview.js`](../configure/index.mdx#configure-story-rendering).

You can define the available background colors using the `values` property:

{/* prettier-ignore-start */}

<CodeSnippets path="addon-backgrounds-values-in-preview.md" />

{/* prettier-ignore-end */}

<Callout variant="warning" icon="🆕">

  **With the [`globals` API](#globals-api)**

>   When using the `globals` API, you must define the available background colors using the `options` property:

Author: I fixed that here: https://github.com/storybookjs/web/pull/212. Are you running on the latest `main` branch?

The Backgrounds addon enables you to change the background color applied to a story by selecting from the list of predefined background colors in the toolbar. If needed, you can force a story to use a specific background color.

You can set the default background color for a story using the `parameters.backgrounds.default` parameter:

{/* prettier-ignore-start */}

<CodeSnippets path="addon-backgrounds-override-parameters-default.md" usesCsf3 />

{/* prettier-ignore-end */}

As implied by the name, this method only sets the default background color for a story. You can still change the background color using the toolbar when viewing the story.

<Callout variant="warning" icon="🆕">

  **With the [`globals` API](#globals-api)**

  When you specify a background color for a story (or a component's stories) using `globals`, the background color is applied and cannot be changed using the toolbar. This is useful when you want to ensure that a story is always rendered on a specific background color.

  {/* prettier-ignore-start */}

  <CodeSnippets path="addon-backgrounds-override-globals.md" usesCsf3 />

  {/* prettier-ignore-end */}

</Callout>

## Extending the configuration

> You can also configure backgrounds per-component or per-story basis through [parameter inheritance](../writing-stories/parameters.mdx#component-parameters).

Reviewer: ```suggestion
You can also configure backgrounds on a per-component or per-story basis through [parameter inheritance](../writing-stories/parameters.mdx#component-parameters).
```

## Configuration

By default, the backgrounds toolbar includes a light and dark background.

But you're not restricted to these backgrounds. You can configure your own set of colors with the `backgrounds` [parameter](../writing-stories/parameters.mdx) in your [`.storybook/preview.js`](../configure/index.mdx#configure-story-rendering). There are two ways to configure the backgrounds:

### Using `options`

<Callout variant="info">
  This API can only be used when the [`backgroundsStoryGlobals` feature flag](../api/main-config/main-config-features.mdx#backgroundsstoryglobals) is enabled.
</Callout>

The available background colors are defined using the `options` property:

{/* prettier-ignore-start */}

<CodeSnippets path="addon-backgrounds-options-in-preview.md" />

{/* prettier-ignore-end */}

### Using `values`

If you're not using the `globals` API, you can define the available background colors using the `values` property:

{/* prettier-ignore-start */}

<CodeSnippets path="addon-backgrounds-values-in-preview.md" />

{/* prettier-ignore-end */}

With either method, available backgrounds can be defined at the project, component, and story levels.

> To apply a default background color, assign the `default` property to one of the available colors.

Author: @ndelangen — Is this API still available when using story globals? I could see value in being able to define a default that is still controllable via the toolbar. But I could also see that complicating the code...

Reviewer 2: The code snippet is the old behavior, which only works if the feature flag is off.

The `default` key is ignored when the feature flag is true.

The opacity of the grid lines.

<Callout variant="warning" icon="🆕">

  **With the [`globals` API](#globals-api)**

  #### `options`

  Type: `Record<string, { name: string; value: string }>`

  Required: See description

  Defines the background colors. Replaces [`values`](#values). See above for a [usage example](#configuration).

>   Required when defining `backgrounds` parameter at project level.

Reviewer: There's a typo here, as demonstrated by my suggestion, and the same issue is present below: users can define this at the component and story levels if they want.
```suggestion
  Required when defining the `backgrounds` parameter at the project level.
```

</Callout>

#### `values`

Type: `{ name: string; value: string }[]`

Required: See description

Available background colors. See above for a [usage example](#configuration).

> Required when defining `backgrounds` parameter at project level.

Reviewer: There's a typo here, should be something like:
```suggestion
Required when defining the `backgrounds` parameter at the project level.
```
Also, this is not entirely true, as documented in the examples above. Users can define this at the component and story level.

Author: > Also, this is not entirely true, as documented in the examples above. Users can define this at the component and story level.

This sentence isn't about when `parameters.backgrounds.values` can be defined. It's about when it _must_ be defined, i.e. when it is required.

Would you find this more clear?

```suggestion
When defining the `backgrounds` parameter at the project level (in `.storybook/preview.js|ts`), you _must_ define the `values` property.
```

Reviewer: That's more accurate.
